File: Data Structures & Algorithms/house-robber-ii/submission-1.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        
        memo = {}

        def dfs(i, robbed_first):
            if i >= len(nums):
                return 0
            if i == len(nums) - 1 and robbed_first:
                return 0
            if (i, robbed_first) in memo:
                return memo[(i, robbed_first)]
            
            if i == 0:
                rob = nums[i] + dfs(i + 2, True)
                skip = dfs(i + 1, False)
                return max(rob, skip)
            else:
                rob = nums[i] + dfs(i + 2, robbed_first)
                skip = dfs(i + 1, robbed_first)
                memo[(i, robbed_first)] = max(rob, skip)
                return max(rob, skip)
        
        return dfs(0, False)

File: Data Structures & Algorithms/combination-target-sum/submission-1.java
class Solution {

    List<List<Integer>> res = new ArrayList<>();
    Deque<Integer> path = new ArrayDeque<>();
    int target;
    int[] nums;

    public List<List<Integer>> combinationSum(int[] nums, int target) {
        this.target = target;
        this.nums = nums;

        dfs(0, 0);
        return res;
    }

    void dfs(int i, int sum){
        if(i >= nums.length || sum > target){
            return;
        }
        if (sum == target){
            res.add(new ArrayList<>(path));
            return;
        }

        // two options, skip or add curr 
        dfs(i + 1, sum);

        sum += nums[i];
        path.offerLast(nums[i]);
        dfs(i, sum);
        //backtrack
        sum -= nums[i];
        path.pollLast();

    }


}
